File: C210/ga-java/src/br/inatel/c210/ga/entity/Problem.java
package br.inatel.c210.ga.entity;

import java.util.List;
import java.util.stream.Collectors;

public class Problem
{
    public static double f(int x, int y)
    {
        return Math.abs(x * y * Math.sin(Math.PI * y / 4.0));
    }

    public static double f(Chromossome c)
    {
        Integer[] data = Chromossome.getFenotype(c.getGenes());
        return f(data[0], data[1]);
    }

    public static double g(int x, int y)
    {
        return f(x, y) + 1;
    }

    public static double g(Chromossome c)
    {
        Integer[] data = Chromossome.getFenotype(c.getGenes());
        return g(data[0], data[1]);
    }

    public static double f_average(List<Chromossome> pop)
    {
        return pop.stream().collect(Collectors.averagingDouble(c -> f(c)));
    }

    public static double g_average(List<Chromossome> pop)
    {
        return pop.stream().collect(Collectors.averagingDouble(c -> g(c)));
    }
}

File: C210/ga-python/modules/problem.py
import math

from .chromossome import Chromossome

class Problem:

    @staticmethod
    def f(x, y):
        return x**2 + y**2

    @staticmethod
    def f_chromossome(chromossome):
        x, y = Chromossome.get_fenotype(chromossome.get_genes())
        return Problem.f(x, y)

    @staticmethod
    def g(x, y):
        return 1 / Problem.f(x, y)

    @staticmethod
    def g_chromossome(chromossome):
        x, y = Chromossome.get_fenotype(chromossome.get_genes())
        return Problem.g(x, y)

    @staticmethod
    def f_average(population):
        avg = 0
        for chromossome in population:
            avg += Problem.f_chromossome(chromossome)
        avg /= len(population)
        return avg

    @staticmethod
    def g_average(population):
        avg = 0
        for chromossome in population:
            avg += g_chromossome(chromossome)
        avg /= len(population)
        return avg
